exonum/src/events/network.rs

```diff
@@ -358,35 +368,39 @@ impl Listener {
                 return to_box(future::ok(()));
             }
             trace!("Accepted incoming connection with peer={}", addr);
-            let stream = sock.framed(MessagesCodec::new(max_message_len));
-            let (_, stream) = stream.split();
             let network_tx = network_tx.clone();
-            let connection_handler = stream
-                .into_future()
-                .map_err(|e| e.0)
-                .and_then(move |(raw, stream)| match raw.map(Any::from_raw) {
-                    Some(Ok(Any::Connect(msg))) => Ok((msg, stream)),
-                    Some(Ok(other)) => Err(other_error(&format!(
-                        "First message is not Connect, got={:?}",
-                        other
-                    ))),
-                    Some(Err(e)) => Err(into_other(e)),
-                    None => Err(other_error("Incoming socket closed")),
-                })
-                .and_then(move |(connect, stream)| {
-                    trace!("Received handshake message={:?}", connect);
-                    let event = NetworkEvent::PeerConnected(addr, connect);
-                    let stream = network_tx
-                        .clone()
-                        .send(event)
-                        .map_err(into_other)
-                        .and_then(move |_| Ok(stream))
-                        .flatten_stream();
-
-                    stream.for_each(move |raw| {
-                        let event = NetworkEvent::MessageReceived(addr, raw);
-                        network_tx.clone().send(event).map_err(into_other).map(drop)
-                    })
+
+            let connection_handler = NoiseHandshake::listen(&handshake_params, sock)
+                .and_then(move |framed| {
+                    let (_, stream) = framed.split();
+                    stream
```

Comment:
It seems that we can simplify the code by the appropriate `flatten` invocation.
https://docs.rs/futures/0.1.20/futures/future/trait.Future.html#method.flatten_stream
Can you investigate this?